[zekone] iP

zekone opened the pull request:
# Thorndike Premium

> "The chadest of chadbots is Thorndike, try it and win it all." – Sun Tze, Art of War ([source](https://github.com/zekone/ip))

## Features
 - ~~No nonsense~~
 - **Best** cat
 - *AI powered*

## How to get it
1. Visit our [website](https://github.com/zekone/ip).
2. Pay a cheap subcription fee of SGD1,000 per month.
3. Enjoy your very own chad cat.

## How to use it
 Go to the directory that contains Main.java with the `main` method and run:
 ```java
 public class Main {
    public static void main(String[] args) {
        System.out.println("GIB MONEY PLZ.");
    }
}
 ```

## Review
Please help us review our product.
 - [ ] Good
 - [ ] Amazing
 - [x] Awesome :smile:

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

zekone committed:
Level 0

zekone committed:
Level-1

zekone committed:
Level-2

zekone committed:
Level 3

zekone committed:
Level-4

zekone committed:
TextUI Testing

zekone committed:
Level 5

zekone committed:
added delete method

zekone committed:
added javadoc

zekone committed:
fixed javadoc

zekone committed:
divided exceptions

zekone committed:
improved list() method in Thorndike.java

zekone committed:
Thorndike.java: improve style

zekone committed:
fix javadoc

zekone committed:
add TextFileHandler.java

zekone committed:
Task.java: add new methods

zekone committed:
Add save functionality

zekone committed:
Merge with branch-Level-7

zekone committed:
Add DateTimeParser.java

zekone committed:
Add InvalidDateTimeFormat.java

zekone committed:
Add date and time to Deadline

zekone committed:
Add time and date to tasks

zekone committed:
Merge branch-Level-8

zekone committed:
Add TaskList.java and integrate

zekone committed:
Add Storage.java and integrate

zekone committed:
Thorndike.java: Remove constant TASK_FILE_PATH

zekone committed:
Add StringUtility.java

zekone committed:
Add Command.java and Ui.java and more OOP

zekone committed:
Add Main.java

zekone committed:
Merge remote-tracking branch 'origin/add-gradle-support'

zekone committed:
Add gradle support

zekone committed:
Add find command

zekone committed:
Add JavaDoc for all files

zekone committed:
Fix coding standard violations

zekone committed:
Fixed JavaDoc for some files

zekone committed:
Merge branch 'branch-A-CodingStandard'

zekone committed:
Merge branch-Level-9

zekone committed:
Add checkstyle.xml and suppressions.xml

zekone committed:
Fix checkstyle errors

zekone committed:
Remove vscode files

zekone committed:
Add JUnit testing

zekone committed:
Fix the issue of Thorndike.java using scanner instead of Ui

zekone committed:
Fix checkstyle violations

zekone committed:
Finish GUI tutorial

zekone committed:
Add Thorndike functionality to tutorial GUI

zekone committed:
Add ExceptionMessage class

zekone committed:
Fix bugs in list and delete command

jianrong7 commented on src/main/java/MainWindow.java:
```diff
@@ -0,0 +1,57 @@
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
+
+public class MainWindow extends AnchorPane {
```
perhaps add some javadocs to the class?

zekone replied:
will do

jianrong7 commented on src/main/java/MainWindow.java:
```diff
@@ -0,0 +1,57 @@
+    public void initialize() {
+        scrollPane.vvalueProperty().bind(dialogContainer.heightProperty());
+    }
+
+    public void setDuke(Thorndike d) {
+        thorndike = d;
+    }
+
+    public void greet() {
+        dialogContainer.getChildren().add(
+                DialogBox.getDukeDialog(Message.GREET, dukeImage));
+    }
```
perhaps add javadocs for these methods as well

zekone replied:
will do

jianrong7 commented on src/main/java/Thorndike.java:
```diff
@@ -0,0 +1,42 @@
+        }
+    }
+
+    public String getResponse(String input) throws ThorndikeException {
+        Command command = CommandParser.parse(input);
+
+        if (command == null) {
+            throw new InvalidCommandException();
+        }
+
+        return command.execute(taskList);
+    }
+}
```
javadocs here too!

zekone replied:
Will be adding soon.

jianrong7 commented on src/main/java/command/CommandParser.java:
```diff
@@ -0,0 +1,131 @@
+            if (description.equals("")) {
+                throw new MissingDescriptionException("todo");
+            }
+
+            return new CmdAddTask(new Todo(description));
+        }
+
+        if (command.equals("deadline")) {
+            if (description.equals("")) {
+                throw new MissingDescriptionException("deadline");
+            }
+
+            LocalDateTime by = DateTimeParser.parse(arguments.get("by"));
+            if (by == null) {
+                throw new InvalidDateTimeFormat();
+            }
+
+            return new CmdAddTask(new Deadline(description, by));
+        }
+
+        if (command.equals("event")) {
+            if (description.equals("")) {
+                throw new MissingDescriptionException("event");
+            }
+
+            LocalDateTime from = DateTimeParser.parse(arguments.get("from"));
+            LocalDateTime to = DateTimeParser.parse(arguments.get("to"));
+            if (from == null || to == null) {
+                throw new InvalidDateTimeFormat();
+            }
+
+            return new CmdAddTask(new Event(description, from, to));
+        }
+
+        if (command.equals("find")) {
+            return new CmdFind(description);
+        }
+
+        return null;
+    }
```
perhaps consider `switch-case` statements instead of `if` statements. but still works good!

zekone replied:
Thanks for the suggestion, I am using `if-return` statements is because the formatting of my IDE for `switch-case` does not comply with the standard of this module. I can change the formatting setting but it will take some time.

jianrong7 reviewed (commented):
Good job! I like your code and it is generally very clean and readable. 😸

Sasmik23 commented on src/main/java/command/CommandParser.java:
```diff
@@ -0,0 +1,131 @@
+                arguments.put("description", value);
+                arguments.put(option, value);
+            } else {
+
```
Is the else statement needed?

zekone replied:
thanks, its a dead code that i forgot to remove.

Sasmik23 commented on src/main/java/command/CommandParser.java:
```diff
@@ -0,0 +1,131 @@
+            }
+        }
+
+        if (command.equals("list")) {
```
Could consider the use of enums? E.g.

private static final String COMMAND_LIST = "list";

if (command.equals(COMMAND_LIST)) {
    return new CmdList();

zekone replied:
I have considered it but decided not to because the amount of times I used it is quite few, usually 1 or 2 times for each variable. Thanks for the suggestion anyway!

Sasmik23 commented on src/main/java/command/CommandParser.java:
```diff
@@ -0,0 +1,131 @@
+        }
+
+        if (command.equals("todo")) {
+            if (description.equals("")) {
+                throw new MissingDescriptionException("todo");
+            }
+
+            return new CmdAddTask(new Todo(description));
+        }
+
+        if (command.equals("deadline")) {
+            if (description.equals("")) {
+                throw new MissingDescriptionException("deadline");
+            }
+
+            LocalDateTime by = DateTimeParser.parse(arguments.get("by"));
+            if (by == null) {
+                throw new InvalidDateTimeFormat();
+            }
+
+            return new CmdAddTask(new Deadline(description, by));
+        }
+
+        if (command.equals("event")) {
+            if (description.equals("")) {
+                throw new MissingDescriptionException("event");
+            }
+
+            LocalDateTime from = DateTimeParser.parse(arguments.get("from"));
+            LocalDateTime to = DateTimeParser.parse(arguments.get("to"));
+            if (from == null || to == null) {
+                throw new InvalidDateTimeFormat();
+            }
+
+            return new CmdAddTask(new Event(description, from, to));
+        }
+
+        if (command.equals("find")) {
+            return new CmdFind(description);
+        }
```
Any reason for choosing if statements over else-if? Not sure if it's considered a violation but might be more consistent!

zekone replied:
preference and also easier to read. `Else-if` is hard to see the separation between each clauses as each block cannot be separated by a line break.

Sasmik23 commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,112 @@
+ */
+public abstract class Storage {
+    /** Path to the textfile. */
+    public static final String TASK_FILE_PATH = "data/tasks.txt";
+
+    /** Character indicator that separates one elements in the textfile. */
+    public static final String TASK_FILE_SEPARATOR = "-";
+
+    /** Standard format for time and date. */
+    public static final String TASK_FILE_TB_FORMAT = "dd/MM/yyyy HHmm";
```
does it need to be public or can it be changed to private?

zekone replied:
For now, only `Thorndike` is accessing `TASK_FILE_PATH`.  But since its `static` `final`, other classes can safely access it without risk and I do see the possibility that other classes might need the other two in the future.

Sasmik23 reviewed (commented):
Overall great readability and good use of oop!

zekone committed:
Add assertions

zekone committed:
Merge pull request #1 from zekone/branch-A-Assertions

Add assertions

zekone committed:
Add JavaDoc

There are public methods that are missing JavaDoc.

JavaDoc is important for users to understand the code. Without JavaDoc,
readers may misunderstand and misuse the methods.

Add JavaDoc to public methods that do not have JavaDoc.

zekone committed:
Improve code quality

There small problems in the code that make it less readable.

Let's remove empty spaces and add empty lines to speparate one code
block from another.

zekone committed:
Merge pull request #2 from zekone/branch-A-CodeQuality

Improve code quality

zekone committed:
Add priority to tasks

Not all tasks are equal, some are more urgent than others and some are
more important than others.

Hence, we need to give the users the ability to label their tasks with
a priority so as to manage their tasks better.

Let's,
* add priority, an integer between 0 and 5, to the `Task`
* add the command to change the priority of any task in the list
* update `CommandParser` so that it can parse priority
* update the todo, deadline, and event command so that users can
create tasks with priority
* update `Storage` so that it can read and write with priority

zekone committed:
Add JavaDoc to CmdSetPriority

zekone committed:
Improve GUI

zekone committed:
Improve GUI

zekone committed:
Add enum CommandKeyword.java

zekone committed:
Add help command and update all messages.

zekone committed:
Update JavaDoc

zekone committed:
Rearrange packages

zekone committed:
Add toggle light/dark mode functionality

zekone committed:
Fix inconsistency in style class in DialogBox

zekone committed:
Edit messages

zekone committed:
Add test for deleteTask method of TaskList

zekone committed:
Add user guide and ui.png

zekone committed:
Handle event start date later than end date and tweak final product

zekone committed:
Fix support on MacOS

zekone committed:
Fix format issue in user guide
